widgets/nuxeo-uploader-behavior.html

```diff
@@ -17,81 +17,88 @@
 
 <script>
   var Nuxeo = Nuxeo || {};
+  (function () {
 
-  /**
-   * @polymerBehavior Nuxeo.UploaderBehavior
-   */
-  Nuxeo.UploaderBehavior = {
-
-    properties: {
-      /**
-       * Accepted file extensions or mime types (comma separated values).
-       */
-      accept: String,
-
-      /**
-       * This flag determines whether the file should be immediately uploaded or not.
-       */
-      immediate: {
-        type: Boolean,
-        value: true
-      },
-
-      /**
-       * Current batch id.
-       */
-      batchId: String,
-
-      /**
-       * List of files in the current batch.
-       */
-      files: {
-        type: Array,
-        value: []
-      },
-
-      /**
-       * Flag that indicates if an upload is in progress.
-       */
-      uploading: {
-        type: Boolean,
-        value: false
-      },
-
-      /**
-       * Allow multiple files to be added to the same batch.
-       */
-      batchAppend: {
-        type: Boolean,
-        value: false
-      }
-    },
-
-    setupDropZone: function(el) {
-      this._dropZone = el;
-      this._dropZone.addEventListener('dragover', this._dragover.bind(this));
-      this._dropZone.addEventListener('dragleave', this._dragleave.bind(this));
-      this._dropZone.addEventListener('drop', this._drop.bind(this));
-    },
-
-    uploadFiles: function(files) {
-      if (!this.accepts(files)) {
-        console.warn('Can only upload ' + this.accept +' files.');
-        return;
-      }
-      if (!this.connection) {
-        throw 'Missing connection';
-      }
+    /**
+     * Initializes a new default upload provider.
+     *
+     * @param {Object} connection an instance of nuxeo-connection to be used
+     * @param {String} accept a string holding a comma separated list of accepted file types or mime types
+     * @param {Boolean} batchAppend a boolean whether files should be appended to the current batch or not 
+     **/
+    function DefaultUploadProvider(connection, accept, batchAppend) {
+      this.connection = connection;
+      this.accept = accept;
+      this.batchAppend = batchAppend;
+      this.uploader = null;
+      this.batchId = null;
+    }
+
+    DefaultUploadProvider.prototype._ensureBatch = function() {
       if (!this.batchAppend || !this.uploader) {
-        this.files = [];
-        this._newBatch().then(this._uploadeFiles.bind(this, files));
+        return this.connection.batchUpload().then(function(uploader) {
+          this.uploader = uploader;
+        }.bind(this));
       } else {
-        this._uploadeFiles(files);
+        return Promise.resolve();
```

Comment:
would it be better to return an error and handle it?

Reply:
No. There's no error here. In this case we just want something **then**able, so that if no new uploader needs to be created, we can just move on and rely on the one already created.